### PR TITLE
spring cloud - custom domain - cli

### PR DESCRIPTION
**Pull request by 1openwindow**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

File: src/spring-cloud/azext_spring_cloud/_params.py [hunk 1]
```diff
@@ -190,3 +190,22 @@ def load_arguments(self, _):
...
         c.argument('deployment', options_list=[
             '--deployment', '-d'], help='Name of an existing deployment of the app. Default to the production deployment if not specified.', validator=validate_deployment_name)
+
+    with self.argument_context('spring-cloud certificate') as c:
+        c.argument('service', service_name_type)
+        c.argument('name', name_type, help='Name of certificate.')
+
+    with self.argument_context('spring-cloud certificate add') as c:
+        c.argument('vault_uri', help='The key vault uri where store the certificate')
+        c.argument('vault_certificate_name', help='The certificate name in key vault')
```
Comment (yungezz) [inline src/spring-cloud/azext_spring_cloud/_params.py]: certificate_name is enough

File: src/spring-cloud/azext_spring_cloud/custom.py [hunk 2]
```diff
@@ -1077,3 +1077,63 @@ def stream(self, amt=2 ** 16, decode_content=None):
...
+    client.certificates.get(resource_group, service, name)
+    return client.certificates.delete(resource_group, service, name)
+
+
+def domain_bind(cmd, client, resource_group, service, app,
+                domain_name,
+                certificate=None):
+    properties = models.CustomDomainProperties()
+    if certificate is not None:
+        certificate_response = client.certificates.get(resource_group, service, certificate)
```
Comment (yungezz) [inline src/spring-cloud/azext_spring_cloud/custom.py]: get result None validation

File: src/spring-cloud/azext_spring_cloud/_params.py [hunk 3]
```diff
@@ -47,8 +47,8 @@ def load_arguments(self, _):
             'is_public', arg_type=get_three_state_flag(), help='If true, assign public domain', default=False)
 
     with self.argument_context('spring-cloud app update') as c:
-        c.argument('is_public', arg_type=get_three_state_flag(),
-                   help='If true, assign public domain')
+        c.argument('is_public', arg_type=get_three_state_flag(),  help='If true, assign public domain')
```
Comment (yucwan) [inline src/spring-cloud/azext_spring_cloud/_params.py]: We are using "endpoint" instead of domain

**Comment by hellokangning** [comment]
Upgrade the `VERSION` in setup.py and history.rst since it would be a feature release

**Comment by yonzhan** [comment]
spring cloud

**Comment by 1openwindow** [comment]
> Upgrade the `VERSION` in setup.py and history.rst since it would be a feature release

updated.

File: src/spring-cloud/azext_spring_cloud/custom.py (hunk 2 above)
Comment (1openwindow) [reply src/spring-cloud/azext_spring_cloud/custom.py:1201]: validate this in rest api

File: src/spring-cloud/azext_spring_cloud/commands.py [hunk 4]
```diff
@@ -86,5 +89,18 @@ def load_command_table(self, _):
         g.custom_command('redis update', 'binding_redis_update')
         g.custom_show_command('remove', 'binding_remove')
 
+    with self.command_group('spring-cloud certificate', client_factory=cf_certificates) as g:
+        g.custom_command('add', 'certificate_add')
+        g.custom_command('show', 'certificate_show', table_transformer=transform_spring_cloud_certificate_output)
```
Comment (qwordy) [inline src/spring-cloud/azext_spring_cloud/commands.py:94]:
```suggestion
        g.custom_show_command('show', 'certificate_show', table_transformer=transform_spring_cloud_certificate_output)

```

File: src/spring-cloud/azext_spring_cloud/_params.py [hunk 5]
```diff
@@ -196,3 +194,22 @@ def load_arguments(self, _):
                    validator=validate_app_name)
         c.argument('deployment', options_list=[
             '--deployment', '-d'], help='Name of an existing deployment of the app. Default to the production deployment if not specified.', validator=validate_deployment_name)
+
+    with self.argument_context('spring-cloud certificate') as c:
+        c.argument('service', service_name_type)
+        c.argument('name', name_type, help='Name of certificate.')
+
+    with self.argument_context('spring-cloud certificate add') as c:
+        c.argument('vault_uri', help='The key vault uri where store the certificate')
```
Comment (qwordy) [inline src/spring-cloud/azext_spring_cloud/_params.py:203]: Can we support name and ID?

Comment (1openwindow) [reply src/spring-cloud/azext_spring_cloud/_params.py:203]: currently not support yet.

Comment (qwordy) [reply src/spring-cloud/azext_spring_cloud/_params.py:203]: I highly suggest support name.

File: src/spring-cloud/azext_spring_cloud/custom.py [hunk 6]
```diff
@@ -1169,3 +1170,63 @@ def stream(self, amt=2 ** 16, decode_content=None):
...
+    )
+    return client.certificates.create_or_update(resource_group, service, name, properties)
+
+
+def certificate_show(cmd, client, resource_group, service, name):
+    return client.certificates.get(resource_group, service, name)
+
+
+def certificate_list(cmd, client, resource_group, service):
+    return client.certificates.list(resource_group, service)
```
Comment (qwordy) [inline src/spring-cloud/azext_spring_cloud/custom.py:1188]: These methods can be omitted. They are default implementation.

**Comment by qwordy** [comment]
Is there any test file?

File: src/spring-cloud/azext_spring_cloud/_help.py [hunk 7]
```diff
@@ -347,3 +347,48 @@
     type: command
     short-summary: Update an Azure Cache for Redis service binding of the app.
 """
+
+helps['spring-cloud certificate add'] = """
```
Comment (qwordy) [inline src/spring-cloud/azext_spring_cloud/_help.py:351]: Add some examples?

Comment (1openwindow) [reply src/spring-cloud/azext_spring_cloud/_help.py:351]: fixed

File: src/spring-cloud/azext_spring_cloud/_help.py [hunk 8]
```diff
@@ -347,3 +347,63 @@
...
+
+helps['spring-cloud certificate list'] = """
+    type: command
+    short-summary: List all certificates in Azure Spring Cloud.
+    examples:
+    - name: List all certificates in spring cloud service.
+      text: az spring-cloud certificate list -o table
+"""
+
+helps['spring-cloud certificate remove'] = """
```
Comment (qwordy) [inline src/spring-cloud/azext_spring_cloud/_help.py:372]:
```
spring-cloud certificate remove
spring-cloud app custom-domain delete
```
I think the verb should be consistent. I prefer "delele"

Comment (yucwan) [reply src/spring-cloud/azext_spring_cloud/_help.py:372]: We should use those verbs in pair right? add/remove, create/delete, bind/unbind

Comment (qwordy) [reply src/spring-cloud/azext_spring_cloud/_help.py:372]: OK.

Comment (yucwan) [reply src/spring-cloud/azext_spring_cloud/_help.py:372]: So I would suggest:
```
spring-cloud certificate remove
spring-cloud app custom-domain unbind
```

File: src/spring-cloud/azext_spring_cloud/tests/latest/test_asc_scenario.py [hunk 9]
```diff
@@ -6,8 +6,57 @@
...
 from azure.cli.testsdk import (ScenarioTest, ResourceGroupPreparer)
 
+# pylint: disable=line-too-long
+# pylint: disable=too-many-lines
 
 TEST_DIR = os.path.abspath(os.path.join(os.path.abspath(__file__), '..'))
+
+@record_only()
+class CustomDomainTests(ScenarioTest):
+    def test_bind_cert_to_domain(self):
```
Comment (qwordy) [inline src/spring-cloud/azext_spring_cloud/tests/latest/test_asc_scenario.py:20]: Add ```@ResourceGroupPreparer(name_prefix='cli_test_bind_cert_to_domain')```

**Comment by qwordy** [comment]
Some checks were not successful

File: src/spring-cloud/azext_spring_cloud/vendored_sdks/appplatform/models/_paged_models.py [hunk 10]
```diff
@@ -51,6 +51,32 @@ class BindingResourcePaged(Paged):
...
+
+    _attribute_map = {
+        'next_link': {'key': 'nextLink', 'type': 'str'},
+        'current_page': {'key': 'value', 'type': '[CertificateResource]'}
+    }
+
+    def __init__(self, *args, **kwargs):
+
+        super(CertificateResourcePaged, self).__init__(*args, **kwargs)
+class CustomDomainResourcePaged(Paged):
```
Comment (qwordy) [inline src/spring-cloud/azext_spring_cloud/vendored_sdks/appplatform/models/_paged_models.py:67]: Lack of blank line

**Pull request merged**